Source/Math/NcclComm.cpp

```diff
@@ -0,0 +1,123 @@
+//
+// Copyright (c) 2016, NVIDIA CORPORATION. All rights reserved.
+// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
+//
+
+#include "NcclComm.h"
+
+#ifdef USE_NCCL
+#include "GPUMatrix.h"
+#include <nccl.h>
+#include <cuda_runtime.h>
+
+namespace Microsoft { namespace MSR { namespace CNTK {
+
+// allows to write cudaFunction() || "error"   (CUDA runtime)
+static void operator||(cudaError_t rc, const char *msg)
+{
+    if (rc != cudaSuccess)
+        RuntimeError("%s: %s (cuda error %d)", msg, cudaGetErrorString(rc), (int) rc);
+}
+
+NcclComm::NcclComm(int deviceId, const MPIWrapperPtr& mpi)
+    : m_ncclComm(nullptr), m_stream(nullptr)
+{
+    if (mpi->IsMultiHost())
+        return;
+
+    size_t numRanks = mpi->NumNodesInUse();
+    MPI_Comm mpiComm = mpi->Communicator();
+    std::vector<int> allDevs(numRanks);
+    MPI_Allgather(&deviceId, 1, MPI_INT, allDevs.data(), 1, MPI_INT, mpiComm)
+        || MpiFail("NcclComm: MPI_Allgather");
+
```

Comment:
should we also check that no two or more workers allocated to the same GPU device?